src/main/c/NativePtr.cpp

```diff
@@ -0,0 +1,352 @@
+  // And clear the ptr without changing refcount
+  vec[0] = 0;
+  vec[1] = 0;
+}
+
+NATIVEMETHOD(void, NativeBase, nativeCopyCtor)(
+  JNIEnv* env,
+  jobject thisObj,
+  jlong b_addrA,
+  jlong b_addrB
+) {
+  if (!isInfoReady) initInfo(env);
+  auto obj = reinterpret_cast<NativeObj*>(b_addrA);
+  // This adds a new reference to the object
+  auto ptr = (obj ? obj->shared_from_this() : NativeObjPtr());
+  initNativePtr(env, thisObj, &ptr);
+}
+
+NATIVEMETHOD(void, NativeBase, copyAssign)(
+  JNIEnv* env,
+  jobject thisObj,
+  jobject srcObj
+) {
+  if (thisObj == srcObj) return;
+  volatile long vecA[2];
+  vecA[0] = env->GetLongField(thisObj, addrAFieldID);
+  vecA[1] = env->GetLongField(thisObj, addrBFieldID);
+  volatile long vecB[2];
+  vecB[0] = env->GetLongField(srcObj, addrAFieldID);
+  vecB[1] = env->GetLongField(srcObj, addrBFieldID);
+  auto ptrA = reinterpret_cast<NativeObjPtr*>(&vecA[0]);
+  auto ptrB = reinterpret_cast<NativeObjPtr*>(&vecB[0]);
+ if (*ptrA != *ptrB) {
+    *ptrA = *ptrB;
+    env->SetLongField(thisObj, addrAFieldID, vecA[0]);
+    env->SetLongField(thisObj, addrBFieldID, vecA[1]);
+  }
+}
+
+NATIVEMETHOD(void, NativeBase, moveAssign)(
+  JNIEnv* env,
+  jobject thisObj,
+  jobject srcObj
+) {
+  if (thisObj == srcObj) return;
+  volatile long vecA[2];
+  vecA[0] = env->GetLongField(thisObj, addrAFieldID);
+  vecA[1] = env->GetLongField(thisObj, addrBFieldID);
+  volatile long vecB[2];
+  vecB[0] = env->GetLongField(srcObj, addrAFieldID);
+  vecB[1] = env->GetLongField(srcObj, addrBFieldID);
+  auto ptrA = reinterpret_cast<NativeObjPtr*>(&vecA[0]);
+  auto ptrB = reinterpret_cast<NativeObjPtr*>(&vecB[0]);
+  *ptrA = std::move(*ptrB);
+  env->SetLongField(thisObj, addrAFieldID, vecA[0]);
+  env->SetLongField(thisObj, addrBFieldID, vecA[1]);
+  env->SetLongField(srcObj, addrAFieldID, vecB[0]);
+  env->SetLongField(srcObj, addrBFieldID, vecB[1]);
+  
+}
+
+NATIVEMETHOD(void, NativeBase, nativeReset)(
+  JNIEnv* env,
+  jobject thisObj,
+  jlong addrA,
+  jlong addrB
+) {
+  volatile long vec[2];
+  vec[0] = addrA;
+  vec[1] = addrB;
+  auto ptr = reinterpret_cast<NativeObjPtr*>(&vec[0]);
+  ptr->reset();
+  // The Scala object fields are cleared in the wrapper
+}
+
+NATIVEMETHOD(long, NativeBase, nativeUseCount)(
+  JNIEnv* env,
+  jobject thisObj,
+  jlong addrA,
+  jlong addrB
+) {
+  volatile long vec[2];
+  vec[0] = addrA;
+  vec[1] = addrB;
+  auto ptr = reinterpret_cast<NativeObjPtr*>(&vec[0]);
+  return ptr->use_count();
+}
+
+//
+// We have constructors corresponding to std::make_shared<T>(...)
+// with various numbers of long arguments.
+//
+
+NATIVEMETHOD(void, NativePtr, nativePtrFuncL0)(
+  JNIEnv* env,
+  jobject thisObj,
+  jlong funcObjAddr
+) {
+  NativeObjPtr ptr = getPtrFuncN(funcObjAddr)();
+  initNativePtr(env, thisObj, &ptr);
+}
+
+NATIVEMETHOD(void, NativePtr, nativePtrFuncL1)(
+  JNIEnv* env,
+  jobject thisObj,
+  jlong funcObjAddr,
+  jlong a0
+) {
+  FILE* f = fopen("/tmp/a.log", "w");
```

Comment:
You left in logging.

Reply:
(Although we should probably make a plan about logging from C++ generally.)

Reply:
Yes - when the C++ is running under downcall from Scala, it could make an upcall to the regular logging mechanism.  But that's not the only possible situation, so I postponed design of that.